src/NumField/Subfields.jl

```diff
@@ -120,38 +123,72 @@ function _subfield_primitive_element_from_basis(K::S, as::Vector{T}) where {
   end
 end
 
-#As above, but for AbsSimpleNumField type
-#In this case, we can use block system to find if an element is primitive.
+function block_system(t::AbsSimpleNumFieldElem, C#=::qAdicConj=#)
+  @assert C isa qAdicConj
+  pr = 1
+  while true
+    c = conjugates(t, C, pr)::Vector{QadicFieldElem}
+    D = Dict{QadicFieldElem, Vector{Int}}()
+    for i = 1:length(c)
+      if haskey(D, c[i])
+        push!(D[c[i]], i)
+      else
+        D[c[i]] = [i]
+      end
+    end
+    bs = sort(collect(values(D)), lt = (a,b) -> isless(a[1], b[1]))
+    if length(bs) * length(bs[1]) == length(c) &&
+        all(x->length(x) == length(bs[1]), bs)
+      return bs
+    end
+    pr *= 2
+    if pr > 100
+      error("probably bad")
+    end
+  end
+end
+
+# As above, but for AbsSimpleNumField type
+# In this case, we can use a block system to find if an element is
+# primitive (and find a primitive element) Input does not need to be
+# a basis, generators are sufficient
 function _subfield_primitive_element_from_basis(K::AbsSimpleNumField, as::Vector{AbsSimpleNumFieldElem})
   if isempty(as) || degree(K) == 1
     return gen(K)
   end
 
+  as = AbsSimpleNumFieldElem[x for x in as if !iszero(x)]
+
   dsubfield = length(as)
 
   @vprintln :Subfields 1 "Sieving for primitive elements"
   # First check basis elements
-  @vprintln :Subfields 1 "Sieving for primitive elements"
-  # First check basis elements
-  Zx = polynomial_ring(ZZ, "x", cached = false)[1]
+  Zx, = polynomial_ring(ZZ, "x", cached = false)
   f = Zx(K.pol*denominator(K.pol))
-  p, d = _find_prime(ZZPolyRingElem[f])
-  #First, we search for elements that are primitive using block systems
-  F = Nemo.Native.finite_field(p, d, "w", cached = false)[1]
-  Ft = polynomial_ring(F, "t", cached = false)[1]
-  ap = zero(Ft)
-  fit!(ap, degree(K)+1)
-  rt = roots(F, f)
-  indices = Int[]
+  _C = get_attribute(K, :subfield_data)
+  if _C === nothing
+    p, d = _find_prime(ZZPolyRingElem[f])
+    _C = qAdicConj(K, p; splitting_field = true)
+    set_attribute!(K, :subfield_data => _C)
+  end
+  C = _C::qAdicConj
+  # First, we search for elements that are primitive using block systems
+  # TODO: use p-adic roots ala Oscar/experimental/Galois.../src/Subfield.jl
+  # prob: get bounds without SLP and GaloisCtx
+  b = Vector{Int}[collect(1:degree(f))] #block system for QQ
+  all_b = Vector{Vector{Int}}[]
   for i = 1:length(as)
-    b = _block(as[i], rt, ap)
-    if length(b) == dsubfield
-      push!(indices, i)
-    end
+    _b = block_system(as[i], C)
+    push!(all_b, _b)
+    b = Vector{Int}[intersect(x, y) for x in b for y in _b]
+    b = Vector{Int}[x for x in b if length(x) > 0]
```

Comment:
is this this used?

Reply:
in line 188 and such to see if the input already has prim. elements
It should be used more, namely in the search for a pe as a comb. of input. Not used if subfield from block is used